src/ngResource/resource.js

```diff
@@ -68,97 +71,111 @@ function parseTemplateParams(template) {
   return urlParams;
 }
 
-var $composeResourceUrlFactory = ['$httpParamSerializer', function($httpParamSerializer) {
-  var encodeUriQuery = angular.$$encodeUriQuery,
-      encodeUriSegment = angular.$$encodeUriSegment;
+function composeResourceUrl(template, params, stripTrailingSlashes) {
+  var protocolAndIpv6 = '',
+      templateParams = parseTemplateParams(template),
+      url = template,
+      queryParams = Object.create(null);
+
+  params = params || Object.create(null);
+
+  // stripTrailingSlashes defaults to true
+  stripTrailingSlashes = (stripTrailingSlashes === undefined) || stripTrailingSlashes;
+
+  // unescape colons
+  url = url.replace(ESCAPED_COLONS_REGEX, ':');
+
+  // extract the protocol
+  url = url.replace(PROTOCOL_AND_IPV6_REGEX, function(match) {
+    protocolAndIpv6 = match;
+    return '';
+  });
+
+  // Inject the param values into the template
+  forEach(templateParams, function(paramInfo, paramName) {
+    var encodedVal,
+        val = params.hasOwnProperty(paramName) ? params[paramName] : undefined;
+
+    if (isDefined(val) && val !== null) {
+      // Encode the param value according to its position in the URL
+      encodedVal = paramInfo.isQueryParamValue ? encodeUriQuery(val, true) : encodeUriSegment(val);
+      // Replace the templateParam segment with the matching param value
+      url = url.replace(new RegExp(':' + paramName + '(\\W|$)', 'g'), function(match, p1) {
+        return encodedVal + p1;
+      });
+    } else {
+      // Collapse segments that map to `undefined` params
+      url = url.replace(new RegExp('(/?):' + paramName + '(\\W|$)', 'g'), function(match,
+          leadingSlashes, tail) {
+        if (tail.charAt(0) === '/') {
+          return tail;
+        } else {
+          return leadingSlashes + tail;
+        }
+      });
+    }
+  });
+
+  // strip trailing slashes and set the url (unless this behavior is specifically disabled)
+  if (stripTrailingSlashes) {
+    url = url.replace(/\/+$/, '') || '/';
+  }
+
+  // Collapse `/.` if found in the last URL path segment before the query.
+  // E.g. `http://url.com/id/.format?q=x` becomes `http://url.com/id.format?q=x`.
+  url = url.replace(/\/\.(?=\w+($|\?))/, '.');
+
+  // Replace escaped `/\.` with `/.`.
+  // (If `\.` comes from a param value, it will be encoded as `%5C.`.)
+  url = protocolAndIpv6 + url.replace(/\/(\\|%5C)\./, '/.');
+
+  // Extract the query params. (Delegate param encoding to $http.)
+  forEach(params, function(value, key) {
+    if (!templateParams[key]) {
+      queryParams[key] = value;
+    }
+  });
+
+  return { url: url, params: queryParams };
+}
+
+var $composeResourceUrlFactory = ['$http', function($http) {
+  var defaultParamSerializer = $http.defaults && $http.defaults.paramSerializer;
```

Comment:
Do we need to read this on every call to `$composeResourceUrl`?

Reply:
Only if we want to do it right 😛